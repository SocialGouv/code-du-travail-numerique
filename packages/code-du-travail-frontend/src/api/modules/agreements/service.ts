import { SearchResponse, ElasticSearchItem, Agreement } from "cdtn-types";
import { elasticsearchClient, elasticIndex, NotFoundError } from "../../utils";
import {
  getAllAgreementsWithContributions,
  getAgreementsBySlugs,
  getAgreementBySlugBody,
  getAgreementsByIds,
} from "./queries";

export const getAllAgreements = async (): Promise<Agreement[]> => {
  const body = getAllAgreementsWithContributions();

  const response = await elasticsearchClient.search<SearchResponse<Agreement>>({
    body,
    index: elasticIndex,
  });

  return response.body.hits.hits
    .map(({ _source }) => _source)
    .sort(orderByAlphaAndMetalurgieLast);
};

export const getBySlugsAgreements = async (
  slugs: string[]
): Promise<ElasticSearchItem[]> => {
  const body = getAgreementsBySlugs(slugs);
  const response = await elasticsearchClient.search({
    body,
    index: elasticIndex,
  });
  return response.body.hits.total.value > 0
    ? response.body.hits.hits.map(({ _source }) => _source)
    : [];
};

export const getByIdsAgreements = async (
  ids: string[]
): Promise<ElasticSearchItem[]> => {
  const body = getAgreementsByIds(ids);
  const response = await elasticsearchClient.search({
    body,
    index: elasticIndex,
  });
  return response.body.hits.total.value > 0
    ? response.body.hits.hits.map(({ _source }) => _source)
    : [];
};

export const getBySlugAgreements = async (slug: string) => {
  const body = await getAgreementBySlugBody(slug);

  const response = await elasticsearchClient.search({
    body,
    index: elasticIndex,
  });
  if (response.body.hits.total.value === 0) {
    throw new NotFoundError({
      message: `Agreement not found, no agreement match ${slug}y`,
      name: "AGREEMENT_NOT_FOUND",
      cause: null,
    });
  }

  return { ...response.body.hits.hits[0]._source };
};

const orderByAlphaAndMetalurgieLast = (a, b) => {
  if (a.url && !b.url) {
    return -1;
  }
  if (!a.url && b.url) {
    return 1;
  }
  if (a.title < b.title) {
    return -1;
  }
  if (a.title > b.title) {
    return 1;
  }

  return 0;
};
