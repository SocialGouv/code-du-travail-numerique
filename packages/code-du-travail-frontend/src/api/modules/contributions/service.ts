import { ElasticSearchItem } from "@socialgouv/cdtn-utils";
import {
  elasticsearchClient,
  elasticDocumentsIndex,
  NotFoundError,
} from "../../utils";
import {
  getAllGenericsContributions,
  getContributionsBySlugs,
  getAllContributionBySlug,
  getContributionsByIds,
} from "./queries";

export const getGenericContributionsGroupByThemes = async () => {
  const body = getAllGenericsContributions();

  const response = await elasticsearchClient.search({
    body,
    index: elasticDocumentsIndex,
  });
  return response.body.hits.hits
    .map(({ _source }) => _source)
    .map((contrib) => {
      contrib.theme = contrib.breadcrumbs[0].label;
      return contrib;
    })
    .reduce(groupByThemes, {});
};

export const getGenericsContributions = async () => {
  const body = getAllGenericsContributions();

  const response = await elasticsearchClient.search({
    body,
    index: elasticDocumentsIndex,
  });
  return response.body.hits.hits.map(({ _source }) => _source);
};

export const getBySlugsContributions = async (
  slugs: string[]
): Promise<ElasticSearchItem[]> => {
  const body = getContributionsBySlugs(slugs);
  const response = await elasticsearchClient.search({
    body,
    index: elasticDocumentsIndex,
  });
  return response.body.hits.total.value > 0
    ? response.body.hits.hits.map(({ _source }) => _source)
    : [];
};

export const getByIdsContributions = async (
  ids: string[]
): Promise<ElasticSearchItem[]> => {
  const body = getContributionsByIds(ids);
  const response = await elasticsearchClient.search({
    body,
    index: elasticDocumentsIndex,
  });
  return response.body.hits.total.value > 0
    ? response.body.hits.hits.map(({ _source }) => _source)
    : [];
};

const groupByThemes = (acc, item) => {
  if (item.theme in acc) acc[item.theme].push(item);
  else acc[item.theme] = [item];
  return acc;
};
