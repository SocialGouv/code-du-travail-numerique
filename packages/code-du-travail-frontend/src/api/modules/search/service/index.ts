export * from "./prequalified";
export * from "./search";
