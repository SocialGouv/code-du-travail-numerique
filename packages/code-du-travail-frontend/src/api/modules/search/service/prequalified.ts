import { SOURCES } from "@socialgouv/cdtn-utils";
const fuzz = require("fuzzball");
import deburr from "lodash.deburr";
import memoizee from "memoizee";
import { elasticDocumentsIndex, elasticsearchClient } from "../../../utils";

const THRESHOLD = 90;
const NON_FUZZY_TOKENS = new Set(["cdd", "cdi", "csp"]);

// Preprocess query : remove accentuation and
//  non alpha numerical to speed up the lookup.
const preprocess = (q) => {
  return fuzz.full_process(deburr(q.trim()));
};

// We populate the saved queries in an object in order to ease
//  searches and map variant matches with the actual known query.

const prequalifiedQuery = {
  bool: {
    filter: [
      { term: { source: SOURCES.PREQUALIFIED } },
      { term: { isPublished: true } },
    ],
  },
};

async function _getPrequalified() {
  const body = await elasticsearchClient.count({
    body: {
      query: prequalifiedQuery,
    },
    index: elasticDocumentsIndex,
  });
  const count = body.count ?? 10000;

  const response = await elasticsearchClient.search<any>({
    body: {
      query: prequalifiedQuery,
      size: count,
    },
    index: elasticDocumentsIndex,
  });

  if (response.hits.hits.length === 0) {
    return null;
  }

  const knownQueriesSet = response.hits.hits.reduce(
    (queries, { _source: query }) => {
      for (const variant of query.variants) {
        const prepro = preprocess(variant);
        queries[prepro] = query;
      }
      const title = preprocess(query.title);
      queries[title] = query;
      return queries;
    },
    {},
  );

  return {
    allVariants: Object.keys(knownQueriesSet),
    knownQueriesSet,
  };
}

const getPrequalified = memoizee(_getPrequalified, {
  maxAge: 1000 * 5 * 60,
  preFetch: true,
  promise: true,
});

const fuzzOptions = {
  full_process: false,
  limit: 2,
  scorer: fuzz.ratio,
  unsorted: false,
};

const testFuzzyAllowed = (query, match) => {
  // don't apply fuzzy matching to specific terms
  const matchingNonFuzz = match
    .split(/\s+/)
    .filter((token) => NON_FUZZY_TOKENS.has(token));

  // if the matching variant contains non-fuzzy tokens,
  // ensure the query also contains it
  const queryTokens = query.split(/\s+/);
  if (matchingNonFuzz.length > 0) {
    return (
      matchingNonFuzz.filter((token) => queryTokens.includes(token)).length > 0
    );
  } else {
    return true;
  }
};

// Test if a given query fuzzy matches with
//  a known one (and variants).
const testMatch = ({ query, knownQueriesSet, allVariants }) => {
  // preprocess query
  const ppQuery = preprocess(query);

  const results = fuzz.extract(ppQuery, allVariants, fuzzOptions);

  if (results && results.length > 1) {
    const closerMatch = results[0][0];
    const bestScore = results[0][1];

    // ensure match is valid
    if (testFuzzyAllowed(query, closerMatch) && bestScore > THRESHOLD) {
      return knownQueriesSet[closerMatch];
    }
  }
};

// find known query if any
export const getPrequalifiedResults = async (query) => {
  const { knownQueriesSet, allVariants } = (await getPrequalified()) as any;

  // allow
  const knownQuery =
    query.length >= 3 && testMatch({ allVariants, knownQueriesSet, query });

  if (knownQuery && knownQuery.refs && knownQuery.refs.length > 0) {
    // get ES results for a known query
    return knownQuery.refs.map((ref) => ({
      _source: ref,
    }));
  }
  return false;
};
