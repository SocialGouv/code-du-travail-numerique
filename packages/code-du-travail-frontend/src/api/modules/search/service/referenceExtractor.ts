// FIXME : borrowed from fiche MT data package, we should extract it from the data package

/*
Extracting references is done in several steps :
1) classifyTokens : we identify valid article references (start with l/r/d then token of shape like 1234-12) => split text into sequence of tokens and give a label to each token
2) identifyCode : we search for the associated code after the ref tokens (valid options are : code du travail / code de la scurite sociale)
3) we group those to constitute structured reference of shape : 
  {
    "article": "L. 2313-8",
    "code": Object {
      "id": "LEGITEXT000006072050",
      "name": "code du travail",
    },
  }
*/

import treebank from "talisman/tokenizers/words/treebank";

const NEGATIVE = "O";
const ARTICLE = "B-ART";
const CODE_PREFIX = "B-COD";
// code du travail
const CODE_TRA = CODE_PREFIX + "_TRA";
// code sécurité sociale
const CODE_SS = CODE_PREFIX + "_SS";
// code any other
const CODE_OTHER = CODE_PREFIX + "_O";

const UNRECOGNIZED = "unrecognized";

const CODE_TRAVAIL = {
  id: "LEGITEXT000006072050",
  name: "code du travail",
};

const CODE_SECU = {
  id: "LEGITEXT000006073189",
  name: "code de la sécurité sociale",
};

const codesFullNames = {
  [CODE_SS]: CODE_SECU,
  [CODE_TRA]: CODE_TRAVAIL,
};

// maximum distance between code tokens and corresponding article ref
const range = 20;

const articleRegEx = new RegExp("^(\\d{1,4}(-\\d+){0,3})\\b"); //          nums        123 123-45 123-45-6 123-45-6-7
export function articleMatcher(token) {
  return token.match(articleRegEx);
}

const validPrefix = ["l", "r", "d"];

// returns :
// 0 if not matching
// 1 if matching prefix only (L.)
// 2 if matching prefix and valid ref (L123.12)
function prefixMatcher(token) {
  const lowToken = token.toLowerCase();

  // if starts with possible prefix
  const matchingPrefix =
    validPrefix.filter((p) => lowToken.startsWith(p)).length > 0;

  if (matchingPrefix) {
    const residual = lowToken.slice(1);

    // case only L
    if (!residual.length) {
      return 1;
    } else {
      // case L.
      if (residual == ".") {
        return 1;
      } else {
        // case L.123-12
        if (residual.slice(0, 1) == "." && articleMatcher(residual.slice(1))) {
          return 2;
        }
        // case L.123-12
        else if (articleMatcher(residual.slice(1))) {
          return 2;
        }
      }
    }
  }
  // no match
  return 0;
}

function infixMatcher(token) {
  // this is quite subtle...
  return ["à", "à"].includes(token);
}

// classify sequence of tokens to identify references to articles
function classifyTokens(tokens) {
  // step 1 : check for prefix matches or articles
  const step1 = tokens.map((token) => {
    const prefix = prefixMatcher(token);
    const infix = infixMatcher(token);
    const article = articleMatcher(token);

    if (prefix > 0) {
      return prefix;
    } else if (article) {
      return 3;
    } else if (infix) {
      return 4;
    } else {
      return 0;
    }
  });

  // step 2 : confirm valid sequences
  // hack : we keep a buffer as last element of the accumulator
  const predictions = step1.reduce(
    (acc, e) => {
      const buffer = acc[acc.length - 1];
      const inSequence = buffer.length > 0;
      const lastElement = buffer[buffer.length - 1];

      // case continue existing
      if (e >= 1 && inSequence) {
        buffer.push(e);
      }

      // case finish existing
      else if (e == 0 && inSequence && lastElement > 1) {
        acc.pop();
        // push buffer
        buffer.forEach(() => acc.push(true));
        // push for current
        acc.push(false);

        acc.push([]);
      }

      // case start (valid start are 1 or 2, as 3 is number only without prefix)
      else if (e > 0 && e < 3 && !inSequence) {
        buffer.push(e);
      }

      // other cases, flush buffer and append current
      else {
        acc.pop();
        acc.push(...buffer.map(() => false));
        acc.push(false);
        acc.push([]);
      }

      return acc;
    },
    [[]]
  );
  // conclude
  const residual = predictions.pop();
  // if ends with bigger than 1, then add residual as true
  if (residual.length > 0 && residual[residual.length - 1] > 1) {
    predictions.push(...residual.map(() => true));
  } else {
    predictions.push(...residual.map(() => false));
  }

  return predictions.map((p) => (p ? ARTICLE : NEGATIVE));
}

function identifyCodes(tokens, predicitions) {
  // we look for "code" tokens (starting a code reference)
  const matchCode = tokens.map((token, i) => {
    return token.toLowerCase() == "code" ? CODE_PREFIX : predicitions[i];
  });

  // we search for entire code references
  const resolvedCodePreds = matchCode.map((pred, i) => {
    if (pred == CODE_PREFIX) {
      const joinedNextTokens = tokens
        .slice(i, i + 5)
        .join(" ")
        .toLowerCase();
      if (joinedNextTokens.startsWith(codesFullNames[CODE_SS].name)) {
        return CODE_SS;
      } else if (joinedNextTokens.startsWith(codesFullNames[CODE_TRA].name)) {
        return CODE_TRA;
      } else {
        return CODE_OTHER;
      }
    } else {
      return pred;
    }
  });

  return resolvedCodePreds;
}

// extract references from free text : tokenize and classify
function extractReferences(text): { code: string; text: string }[] {
  const tokens = treebank(text);
  let predictions = classifyTokens(tokens);
  predictions = identifyCodes(tokens, predictions);

  // console.log(tokens);
  // console.log(predictions);

  // group continuous positives tokens and set code
  // while continuous match, merge
  // if code, then associate it to articles within range
  return tokens
    .map((token, index) => {
      return { index, pred: predictions[index], token };
    })
    .reduce((acc, { token, index, pred }) => {
      // case article : we start or merge
      if (pred == ARTICLE) {
        if (acc.length == 0) {
          acc.push({ index, token });
        } else {
          const last = acc[acc.length - 1];
          // case continuous : we merge
          if (last.index + 1 == index) {
            last.token = `${last.token} ${token}`;
            last.index = index;
          } else {
            acc.push({ index, token });
          }
        }
      }
      // case code, we associate it to articles within range
      else if (pred.startsWith(CODE_PREFIX) && acc.length > 0) {
        acc.forEach((match) => {
          // if no code yet and in range
          if (!match.code && match.index + range >= index) {
            if (pred in codesFullNames) {
              match.code = codesFullNames[pred];
            } else {
              match.code = UNRECOGNIZED;
            }
          }
        });
      }

      return acc;
    }, [])
    .filter(({ code }) => {
      // valid cases are no code or code different than UNRECOGNIZED (for other codes : rural, education...)
      return !code || (code && code != UNRECOGNIZED);
    })
    .map(({ token, code }) => {
      return { code, text: token };
    });
}

export {
  classifyTokens,
  CODE_SECU,
  CODE_TRAVAIL,
  codesFullNames,
  extractReferences,
};
