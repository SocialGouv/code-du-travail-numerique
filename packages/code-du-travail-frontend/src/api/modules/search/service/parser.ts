import { token } from "@styled-system/tokens";
import { getIdccByQuery } from "../../idcc";
import {
  articleMatcher,
  classifyTokens,
  extractReferences,
} from "./referenceExtractor";
import { extractHits } from "./search";
import { complex } from "talisman/stemmers/french/unine";
import fingerprint from "talisman/tokenizers/fingerprint";

const keyword_std = [
  "a conservatoir mise pied",
  "a convoc disciplinair entretien prealabl sanction une",
  "a de depart indemnit la retrait",
  "a de depart la prim retrait",
  "a lamiabl ruptur",
  "a mise pied",
  "acident de trajet",
  "acord dentrepris",
  "acquis aret cong malad paye pendant",
  "acquis cong paye",
  "afichag obligatoir",
  "alait",
  "alternanc contrat daprentisag et",
  "amenag de du post travail",
  "ancienet",
  "aprentisag",
  "aret de maintien malad salair",
  "aret de travail",
  "aret malad",
  "atest de travail",
  "au avenant contrat de travail",
  "au harcel moral travail",
  "au inaptitud travail",
  "averti",
  "calcul compt de sold tout",
  "carenc cdd de delai",
  "cdd contrat",
  "cdd renouvel",
  "cdi convention ruptur",
  "cdi demision",
  "cdi preavi",
  "certificat de travail",
  "claus concurenc de non",
  "compens repo",
  "complementair heur",
  "compt de sold tout",
  "cong dancienet",
  "cong dece",
  "cong dece grand parent pour",
  "cong enfant malad",
  "cong et fraction paye",
  "cong et malad paye",
  "cong et mi paye temp therapeut",
  "cong even familial les pour",
  "cong exception",
  "cong parental",
  "cong paternit",
  "cong san sold",
  "contrat de delai document fin",
  "cse",
  "cse entrepris",
  "csp",
  "dan du le paie retard salair",
  "dancienet prim",
  "de delai desai et period prevenanc ruptur",
  "de delai desai period prevenanc",
  "de deleg heur",
  "de demision du dure preavi",
  "de demision letr",
  "de deplac domicil frai trajet travail",
  "de et maintien malad salair",
  "de fich paie",
  "de frai indemnit repa",
  "de gril salair",
  "de indemnit licenc",
  "de letr licenc model type",
  "de paus temp",
  "dece employeu",
  "demenag",
  "dimanch le travail",
  "disciplinair pouvoi",
  "du dure preavi",
  "enfant malad",
  "et mutuel retrait",
  "faut licenc pour",
  "fer jour travail",
  "harcel moral",
  "heur non paye suplementair",
  "impay salair",
  "indemnit licenc",
  "indemnit repa",
  "medical travail visit",
  "minimum salair",
  "mitemp therapeut",
  "preavi",
  "prevoyanc",
  "retard",
  "syntec",
];

export const CC_SCORE_THRESHOLD = 20;

export enum LabelEnum {
  CC = "cc",
  CC_FOUND = "cc_found",
  KEYWORD = "keyword",
  KEYWORD_STD = "keyword_standard",
  ARTICLE = "article",
  THEME = "theme",
  NATURAL = "natural",
}

export const LABELS = Object.values(LabelEnum);

export type StructuredQueryLabel = {
  label: LabelEnum;
  details?: string;
  score?: number;
};

const matchingCC = [
  "convention",
  "collective",
  "collectif",
  "national",
  "branche",
  "idcc",
  "ccn",
];

// const getTokens()
// const tokenDiff

// remove

const ccSearch = async (query): Promise<StructuredQueryLabel | undefined> => {
  const idccResults = await getIdccByQuery(query, 1);

  const hits = extractHits(idccResults);

  //   console.log({ query });
  //   console.log({ hits });

  if (hits.length > 0 && (hits[0]["_score"] || 0) > CC_SCORE_THRESHOLD) {
    const hit = hits[0];
    return {
      label: LabelEnum.CC_FOUND,
      details: hit["_source"]["slug"],
      score: hit["_score"] || undefined,
    };
  }
};

const isNumToken = (token) =>
  // TODO make use of a curated list of idcc valid codes
  !isNaN(token) && !isNaN(parseFloat(token)) && parseFloat(token) > 10;

const isCC = async (query: string) => {
  const tokens = fingerprint(query);

  if (
    matchingCC.find((t) => tokens.includes(t))?.length ||
    tokens.find(isNumToken)
  ) {
    const remainingTokens = Array.from([
      ...tokens.filter((t) => !matchingCC.includes(t)),
    ]);

    // in case we find a number, we assume idcc search and remove other tokens to avoid noise
    // otherwise, we run search with all non-matchingCC tokens
    const searchTokens = remainingTokens.find(isNumToken)
      ? remainingTokens.filter(isNumToken)
      : remainingTokens;

    // TODO remove "theme"-like tokens when performing CC search
    // or only select tokens that appear in all conventions titles

    const found = await ccSearch(searchTokens.join(" "));
    return (
      found ||
      (tokens.find(isNumToken)
        ? undefined
        : {
            label: LabelEnum.CC,
            details: `cc related content, but no match ${searchTokens.join(" ")}`,
          })
    );
  } else {
    const foundNoCC = ccSearch(query);
    return foundNoCC;
  }
};

const getStandardKeyword = (pQuery: string[]) =>
  keyword_std.includes(pQuery.join(" "))
    ? { label: LabelEnum.KEYWORD_STD }
    : undefined;

const getArticles = (query): StructuredQueryLabel[] | undefined => {
  // proper references
  let refs = extractReferences(query);
  if (refs.length < 1) {
    // first token matching regex
    const match = articleMatcher(query.split(" ")[0]);
    if (match) {
      refs.push(match);
    }
  }
  return refs.length > 0
    ? refs.map((r) => ({ label: LabelEnum.ARTICLE, details: r.text }))
    : undefined;
};

const prepro = (token: string): string[] => fingerprint(token).map(complex);

const getThemes = (pQuery: string[], themes: string[]) => {
  //   console.log({ query, themes });

  const pThemes = themes.map((theme) => ({ theme, pTheme: prepro(theme) }));

  let match = pThemes.find((th) => th.pTheme == pQuery);
  if (!match) {
    // in this case, we should select the highest one in the hierarchy
    match = pThemes.find((th) => pQuery.every((to) => th.pTheme.includes(to)));
  }
  return match ? { label: LabelEnum.THEME, details: match.theme } : undefined;
};

const isNatural = (query: string) => {
  const tokens = fingerprint(query);
  const matching = [
    "comment",
    "combien",
    "quand",
    "je",
    "nous",
    "me",
    "il",
    "doit",
    "quel",
    "quels",
    "on",
  ];
  return (
    query.includes("?") ||
    matching.find((m) => tokens.includes(m)) ||
    tokens.length >= 8
  );
};

export const parseQuery = async (query: string, themes: string[]) => {
  const results: StructuredQueryLabel[] = [];

  const pQuery = prepro(query);

  const cc = await isCC(query);
  if (cc) {
    results.push(cc);
  }

  const articles = getArticles(query);
  if (articles?.length) {
    results.push(...articles);
  }

  const matchingTheme = getThemes(pQuery, themes);
  if (matchingTheme) {
    results.push(matchingTheme);
  }

  const matchingStd = getStandardKeyword(pQuery);
  if (matchingStd) {
    results.push(matchingStd);
  }

  if (results.length == 0) {
    const natural = isNatural(query);
    results.push({
      label: natural ? LabelEnum.NATURAL : LabelEnum.KEYWORD,
      details: query,
    });
  }

  if (results[0].label == LabelEnum.CC && results.length > 1) {
    results.pop();
  }

  return { nq: pQuery, results };
};
