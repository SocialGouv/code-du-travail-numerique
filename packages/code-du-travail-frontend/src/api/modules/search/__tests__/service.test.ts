import { searchWithQuery } from "../service";

describe("Search", () => {
  it("searchWithQuery", async () => {
    const result = await searchWithQuery("d√©mission", false);
    expect(result).toMatchSnapshot();
  });
});
