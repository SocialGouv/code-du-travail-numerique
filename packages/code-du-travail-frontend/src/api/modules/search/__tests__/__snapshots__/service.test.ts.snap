// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search searchWithQuery 1`] = `
{
  "articles": [
    {
      "_score": undefined,
      "algo": "pre-qualified",
      "breadcrumbs": [],
      "cdtnId": "6638a02755",
      "description": "Le contrat de travail à durée indéterminée peut être rompu à l'initiative de l'employeur ou du salarié, ou d'un commun accord, dans les conditions prévues par les dispositions du présent titre. Ces dispositions ne sont pas applicables pendant la période d'essai",
      "slug": "l1231-1",
      "source": "code_du_travail",
      "title": "L1231-1",
    },
    {
      "_score": undefined,
      "algo": "pre-qualified",
      "breadcrumbs": [],
      "cdtnId": "ea03713a82",
      "description": "En cas de démission, l'existence et la durée du préavis sont fixées par la loi, ou par convention ou accord collectif de travail. En l'absence de dispositions légales, de convention ou accord collectif de travail relatifs au préavis, son existence et sa durée résultent des usages pratiqués dans la localité et dans la profession. Un décret en Conseil d'Etat détermine les modalités d'application du présent article",
      "slug": "l1237-1",
      "source": "code_du_travail",
      "title": "L1237-1",
    },
    {
      "_score": undefined,
      "algo": "pre-qualified",
      "breadcrumbs": [],
      "cdtnId": "4c53e81904",
      "description": "La rupture d'un contrat de travail à durée indéterminée à l'initiative du salarié ouvre droit, si elle est abusive, à des dommages et intérêts pour l'employeur. En cas de litige, le juge se prononce conformément aux dispositions de l'article L. 1235-1 . Un décret en Conseil d'Etat détermine les modalités d'application du présent article",
      "slug": "l1237-2",
      "source": "code_du_travail",
      "title": "L1237-2",
    },
    {
      "_score": 3.6866884,
      "cdtnId": "f13147ec33",
      "description": "Le salarié informe son employeur de sa démission, en application de l'article L. 1225-66 , par lettre recommandée avec avis de réception ou remise contre récépissé. Il adresse à l'employeur sa demande de réembauche, en application de l'article L. 1225-67 , par lettre recommandée avec avis de réception ou remise contre récépissé",
      "slug": "r1225-18",
      "source": "code_du_travail",
      "title": "R1225-18",
      "url": "https://www.legifrance.gouv.fr/affichCodeArticle.do;?idArticle=LEGIARTI000018537778&cidTexte=LEGITEXT000006072050",
    },
  ],
  "documents": [
    {
      "_score": undefined,
      "algo": "pre-qualified",
      "breadcrumbs": [
        {
          "label": "Départ de l’entreprise",
          "position": 8,
          "slug": "/themes/depart-de-lentreprise",
        },
        {
          "label": "Démission",
          "position": 1,
          "slug": "/themes/demission",
        },
      ],
      "cdtnId": "8c42435f9a",
      "description": "La démission permet au salarié, à son initiative, de rompre son contrat de travail, sous conditions.",
      "slug": "demission-dun-salarie",
      "source": "fiches_service_public",
      "title": "Démission d'un salarié",
    },
    {
      "_score": undefined,
      "algo": "pre-qualified",
      "breadcrumbs": [
        {
          "label": "Départ de l’entreprise",
          "position": 8,
          "slug": "/themes/depart-de-lentreprise",
        },
        {
          "label": "Démission",
          "position": 1,
          "slug": "/themes/demission",
        },
      ],
      "cdtnId": "5d4a0ee541",
      "description": "La démission permet au salarié de rompre son CDI de sa propre initiative. Si un délai de préavis est prévu, il doit être respecté.",
      "slug": "la-demission",
      "source": "page_fiche_ministere_travail",
      "title": "La démission",
    },
    {
      "_score": undefined,
      "algo": "pre-qualified",
      "breadcrumbs": [
        {
          "label": "Départ de l’entreprise",
          "position": 8,
          "slug": "/themes/depart-de-lentreprise",
        },
        {
          "label": "Démission",
          "position": 1,
          "slug": "/themes/demission",
        },
      ],
      "cdtnId": "4e7e702fb5",
      "description": "Le salarié qui démissionne n'a pas droit aux allocations chômage (ARE), sauf cas particuliers (démission légitime, reliquat de droits...).",
      "slug": "un-salarie-peut-il-percevoir-lallocation-chomage-en-cas-de-demission",
      "source": "fiches_service_public",
      "title": "Un salarié peut-il percevoir l'allocation chômage en cas de démission ?",
    },
    {
      "_score": undefined,
      "algo": "pre-qualified",
      "breadcrumbs": [
        {
          "label": "Congés et repos",
          "position": 4,
          "slug": "/themes/conges-et-repos",
        },
        {
          "label": "Congés",
          "position": 1,
          "slug": "/themes/conges",
        },
        {
          "label": "Congés payés",
          "position": 0,
          "slug": "/themes/conges-payes",
        },
      ],
      "cdtnId": "52562fc16f",
      "description": "Le salarié peut démissionner pendant ses congés payés, mais le préavis commence à courir à l'issue de ses congés (sauf accord avec l'employeur).",
      "slug": "peut-on-demissionner-pendant-ses-conges-payes",
      "source": "fiches_service_public",
      "title": "Peut-on démissionner pendant ses congés payés ?",
    },
    {
      "_score": undefined,
      "algo": "pre-qualified",
      "breadcrumbs": [
        {
          "label": "Départ de l’entreprise",
          "position": 8,
          "slug": "/themes/depart-de-lentreprise",
        },
        {
          "label": "Démission",
          "position": 1,
          "slug": "/themes/demission",
        },
      ],
      "cdtnId": "6b17d70cd9",
      "description": "Le salarié peut démissionner pendant un arrêt maladie, sauf s'il est en période d'essai.",
      "slug": "peut-on-demissionner-pendant-un-arret-maladie",
      "source": "fiches_service_public",
      "title": "Peut-on démissionner pendant un arrêt maladie ?",
    },
    {
      "_score": undefined,
      "algo": "pre-qualified",
      "breadcrumbs": [
        {
          "label": "Départ de l’entreprise",
          "position": 8,
          "slug": "/themes/depart-de-lentreprise",
        },
        {
          "label": "Démission",
          "position": 1,
          "slug": "/themes/demission",
        },
      ],
      "cdtnId": "184ebe1ef6",
      "description": "Oui. Toute salariée peut démissionner pendant un congé maternité, dans les conditions habituelles prévues en matière de démission.",
      "slug": "peut-on-demissionner-pendant-un-conge-maternite",
      "source": "fiches_service_public",
      "title": "Peut-on démissionner pendant un congé maternité ?",
    },
    {
      "_score": undefined,
      "algo": "pre-qualified",
      "breadcrumbs": [
        {
          "label": "Départ de l’entreprise",
          "position": 8,
          "slug": "/themes/depart-de-lentreprise",
        },
        {
          "label": "Démission",
          "position": 1,
          "slug": "/themes/demission",
        },
      ],
      "cdtnId": "503c412884",
      "description": "Le salarié peut démissionner pendant son congé parental, dans le respect de la procédure habituelle prévue en matière de démission.",
      "slug": "peut-on-demissionner-pendant-un-conge-parental",
      "source": "fiches_service_public",
      "title": "Peut-on démissionner pendant un congé parental ?",
    },
    {
      "_score": undefined,
      "algo": "pre-qualified",
      "breadcrumbs": [
        {
          "label": "Départ de l’entreprise",
          "position": 8,
          "slug": "/themes/depart-de-lentreprise",
        },
        {
          "label": "Démission",
          "position": 1,
          "slug": "/themes/demission",
        },
      ],
      "cdtnId": "4e0dedaa2e",
      "description": "Le salarié d'un particulier employeur qui souhaite démissionner doit informer son employeur et respecter un préavis. Il peut bénéficier d'une indemnité compensatrice de préavis et éventuellement d'une indemnité compensatrice de congés payés. L'employeur doit lui remettre des documents de fin de contrat.",
      "slug": "demission-du-salarie-a-domicile-employe-par-un-particulier",
      "source": "fiches_service_public",
      "title": "Démission du salarié à domicile employé par un particulier",
    },
    {
      "_score": undefined,
      "algo": "pre-qualified",
      "breadcrumbs": [
        {
          "label": "Départ de l’entreprise",
          "position": 8,
          "slug": "/themes/depart-de-lentreprise",
        },
        {
          "label": "Démission",
          "position": 1,
          "slug": "/themes/demission",
        },
      ],
      "cdtnId": "38474d168c",
      "description": "L'assistante maternelle (ou assistant maternel) agréée qui souhaite rompre son contrat de travail doit informer le particulier employeur par écrit et respecter un préavis. À la fin de celui-ci, l'employeur remet les documents de fin de contrat.",
      "slug": "demission-dune-assistante-maternelle",
      "source": "fiches_service_public",
      "title": "Démission d'une assistante maternelle",
    },
    {
      "_score": 39.56697,
      "action": "Calculer",
      "algo": "fulltext",
      "breadcrumbs": [
        {
          "label": "Départ de l’entreprise",
          "position": 8,
          "slug": "/themes/depart-de-lentreprise",
        },
        {
          "label": "Démission",
          "position": 1,
          "slug": "/themes/demission",
        },
      ],
      "cdtnId": "1eea193273",
      "description": "Calculez la durée de préavis à respecter en cas de démission",
      "slug": "preavis-demission",
      "source": "outils",
      "title": "Préavis de démission",
    },
    {
      "_score": 36.872955,
      "action": "Calculer",
      "algo": "fulltext",
      "breadcrumbs": [
        {
          "label": "Départ de l’entreprise",
          "position": 8,
          "slug": "/themes/depart-de-lentreprise",
        },
      ],
      "cdtnId": "d8fca710c5",
      "description": "Calculez le nombre d'heures d'absence autorisée pour rechercher un emploi pendant un préavis",
      "slug": "heures-recherche-emploi",
      "source": "outils",
      "title": "Heures d'absence pour rechercher un emploi",
    },
    {
      "_score": 32.03567,
      "algo": "fulltext",
      "breadcrumbs": [
        {
          "label": "Départ de l’entreprise",
          "position": 8,
          "slug": "/themes/depart-de-lentreprise",
        },
        {
          "label": "Documents à remettre au salarié",
          "position": 0,
          "slug": "/themes/documents-a-remettre-au-salarie",
        },
      ],
      "cdtnId": "59a4d4173c",
      "description": "Le certificat de travail est un document remis par l'employeur au salarié à la date de fin du contrat de travail. Il est remis au salarié pour tout type de contrat de travail (CDI, CDD...), quel que soit le motif de la rupture (démission, licenciement...).",
      "slug": "certificat-de-travail",
      "source": "fiches_service_public",
      "title": "Certificat de travail",
      "url": "https://www.service-public.fr/particuliers/vosdroits/F87",
    },
  ],
  "themes": [
    {
      "_score": 2.5448813,
      "algo": "fulltext",
      "cdtnId": "a9ccf1d613",
      "slug": "demission",
      "source": "themes",
      "title": "Démission",
    },
  ],
}
`;
